src/sentry/debug_files/artifact_bundles.py

```diff
@@ -34,14 +50,374 @@
+            pre_serialized_index["files_by_url"] = flat_file_index["files_by_url"]
+
+        if "files_by_debug_id" in flat_file_index:
+            pre_serialized_index["files_by_debug_id"] = flat_file_index["files_by_debug_id"]
+
+        return json.dumps(pre_serialized_index)
+
+    @classmethod
+    def _serialize_bundles(cls, bundles: Bundles) -> List[Dict[str, Union[int, str]]]:
+        serialized_bundles: List[Dict[str, Union[int, str]]] = []
+
+        for bundle in bundles:
+            serialized_bundles.append(
+                {"id": bundle["id"], "timestamp": datetime.isoformat(bundle["timestamp"])}
+            )
+
+        return serialized_bundles
+
+
+T = TypeVar("T")
+
+
+class FlatFileIndex:
+    def __init__(self, store: FlatFileIndexStore):
+        self.store = store
+
+        # By default, a flat file index is empty.
+        self._bundles: Bundles = []
+        self._files_by_url: FilesByUrl = {}
+        self._files_by_debug_id: FilesByDebugID = {}
+
+    @staticmethod
+    def build(store: FlatFileIndexStore) -> Optional[FlatFileIndex]:
+        index = FlatFileIndex(store)
+
+        try:
+            index._build()
+            return index
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_building")
+            return None
+
+    def _build(self) -> None:
+        loaded_index = self.store.load()
+
+        # If we don't find an index, we will silently not mutate the state, since it's initialized as empty.
+        if loaded_index is not None:
+            self._bundles = cast(Bundles, loaded_index.get("bundles", []))
+            self._files_by_url = cast(FilesByUrl, loaded_index.get("files_by_url", {}))
+            self._files_by_debug_id = cast(
+                FilesByDebugID, loaded_index.get("files_by_debug_id", {})
+            )
+
+    def merge(self, artifact_bundle: ArtifactBundle, archive: ArtifactBundleArchive) -> bool:
+        try:
+            # We want to index by url if the store has an identifier which is being indexed by release.
+            self._merge(artifact_bundle, archive, self.store.identifier.is_indexing_by_release())
+            return True
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_merging")
+            return False
+
+    def _merge(
+        self, artifact_bundle: ArtifactBundle, archive: ArtifactBundleArchive, index_urls: bool
+    ):
+        bundle_index = self._update_bundles(artifact_bundle)
+
+        if index_urls:
+            self._iterate_over_urls(bundle_index, archive)
+        else:
+            self._iterate_over_debug_ids(bundle_index, archive)
+
+    def _update_bundles(self, artifact_bundle: ArtifactBundle) -> int:
+        if artifact_bundle.date_last_modified is None:
+            raise Exception("Can't index a bundle with no date last modified")
```

Comment:
The caller can check this precondition as well, no need to pass in the `ArtifactBundle` model here.

Reply:
That was my idea of the `BundleMeta` object. The invariants should be encoded in the type system if it makes sense, and is possible.